### PR TITLE
Week 12 homeworks

### PR DESCRIPTION
**Pull request by JolandaNava**

### CODE REVIEW

File: src/Week12/Risk.hs [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+--------------------------------------------
+-- implementing rollNDie equivalent
+-- using (Rand StdGen) Monad instance
+-- instead of Applicative instance
+rollThreeDie :: Rand StdGen [DieValue]
+rollThreeDie =
+  die >>= \x ->
+  die >>= \y -> 
+  die >>= \z ->
+  return [x,y,z]
```
Comment (chshersh) [inline src/Week12/Risk.hs:92]: You can use `replicateM` function here

File: src/Week12/Risk.hs [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+rollTwoDie :: Rand StdGen [DieValue]
+rollTwoDie =
+  die >>= \x ->
+  die >>= \y -> 
+  return [x,y]
+
+rollOneDice :: Rand StdGen [DieValue]
+rollOneDice =
+  die >>= \x -> 
+  return [x]
```
Comment (chshersh) [inline src/Week12/Risk.hs]: `m >>= \x -> return (f x)` is `f <$> m`. You don't always need monad, sometimes it's enough to use `Functor` instance

File: src/Week12/Risk.hs [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+    isSuccess (Battlefield _ d)
+      | d == 0 = True
+      | otherwise = False 
+
+successProb :: Battlefield -> Rand StdGen Double
+successProb b = (/ 1000.0) . fromIntegral . length . filter id <$> helper 1000
+  where
+    helper :: Int -> Rand StdGen [Bool]
+    helper 0 = return []
+    helper n = (:) <$> suceessfulInvasion b <*> helper (n-1)
```
Comment (chshersh) [inline src/Week12/Risk.hs:129]: This `helper` also looks like `replicateM`

File: src/Week12/Risk.hs [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+        versus :: (DieValue, DieValue) -> (Int,Int)
+        versus (a,d) | a > d = (0, -1)
+        versus _ = (-1,0)
+
+    rollDie :: (Int,Int) -> Rand StdGen ([DieValue], [DieValue])
+    rollDie (a, d) = (,) <$> rollNDie  a <*> rollNDie d
+
+rollNDie :: Int -> Rand StdGen [DieValue]
+rollNDie 0 = return []
+rollNDie n = (:) <$> die <*> rollNDie (n-1)
```
Comment (chshersh) [inline src/Week12/Risk.hs:64]: `replicateM` again 🙂

**Comment by JolandaNava** [comment]
@ChShersh  I've added the implementation of `RollNDie` with `replicateM` but left the other implementations there: it's helpful for me to see how to implement using just the operators.

**Review by chshersh — APPROVED** [review]

**Pull request merged**